File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {

     boolean possible(int mid,int weights[], int days){
        int cap=0;
        int totaldays=1;

        for(int x:weights){
            if(cap+x>mid){
                totaldays++;
                cap=0;
            }
            cap+=x;
            if(totaldays>days) return false;
        }
        return true;
    }

    public int shipWithinDays(int[] weights, int days) {
        int low=Arrays.stream(weights).max().getAsInt();
        int high=Arrays.stream(weights).sum();

        while(low<=high){
            int mid=low+(high-low)/2;
            if(possible(mid,weights,days)) high=mid-1;
            else low=mid+1;
        }
        return low;
    }
}

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cs
public class Solution {

    public bool possible(int mid,int[] weights,int days){
        int daysTaken=1;
        int cap=0;
        foreach(int x in weights){
            if(cap+x<=mid){
                cap+=x;
            }else{
                daysTaken++;
                cap=x;
            }

            if(daysTaken>days) return false;
        }
        return true;
    }

    public int ShipWithinDays(int[] weights, int days) {
        int n=weights.Length;
        int low=weights.Max(),high=weights.Sum();
        while(low<=high){
            int mid=low+(high-low)/2;
            if(possible(mid,weights,days)) high=mid-1;
            else low=mid+1;
        }
        return low;
    }
}